python/pyarrow/util.py

```diff
@@ -41,6 +41,24 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _deprecate_class(old_name, new_class, next_version,
+                     instancecheck=True):
+    """
+    Raise warning if a deprecated class is used in an isinstance check.
```

Comment:
Shouldn't it also raise when instantiated?

Reply:
> Really nice!
> 
> Could we add a `is_valid` attribute to the python scalar as well? Now the only way to check for a null value is to do `.as_py() is None` ?
Definitely!
> 
> Might not be related to the changes in this PR, but reviewing this triggered me to test scalar casting:
> 
> ```
> In [2]: s = pa.scalar(pd.Timestamp("2012-01-01")) 
>    ...:  
>    ...: import pyarrow.compute as pc 
>    ...: pc.cast(s, pa.timestamp('ns'))  
> ../src/arrow/compute/kernels/scalar_cast_temporal.cc:130:  Check failed: (batch[0].kind()) == (Datum::ARRAY) 
> ...
> Aborted (core dumped)
> ```
I think the cast kernel doesn't support scalars yet, on the other hand the scalars have custom `CastTo` implementation which we may want to remove in favor of `compute.Cast`?

cc @bkietz